GlobalNotification -> BannerNotification(s)

bkoval opened the pull request:
This is an attempt to refactor GlobalNotification. Summary:
- Renaming it to BannerNotification
- Moving all the related files to the single extension directory
- Creating an AMD module for BN, but keeping the window global
- Adding support for multiple instances of notifications
- Adding generic method for ajax connection errors.
- Notifications are now objects that can be instantiated with 'new' keyword
- Setters and methods can be chained

PHP Usage

``` php
BannerNotificationController::addNotification(...);
```

instead of 

``` php
NotificationController::addNotification(...);
```

JavaScript Usage

``` javascript
//where possible
require(['BannerNotification'], function(BannerNotification){

    new BannerNotification('MyMessage", 'error').show();

    new BannerNotification().setContent('MyMessage').setType('error').show();

    var notification = new BannerNotification('MyMessage').setType('notify');
    notification.show();
    notification.hide();
});

//Where not possible to use AMD, there's still support for:
new window.BannerNotification().setContent('Hello').setType('confirm').show();
```

Also applied AMD version where it was possible and made WIkiaInYourLang and CookiePolicy use $element exposed by notification instance for binding events.

@kvas-damian @lizlux Please, review and after i address the Code Review notes I will ask owners of all the extensions using notifications to review the code.

I also added a wikia.onScroll module as a global place for binding to scroll event on window:

``` javascript
require(['wikia.onScroll'], function(onScroll){
    function doSomething(){}

    onScroll.bind(doSomething);
    //doSomething is getting triggered on scroll event now

    onScroll.unbind(doSomething);
    //doSomething has been unboud, it's not getting triggered anymore
});
```

bkoval committed:
Moving php files to an extension and registering it

bkoval committed:
Moving js for GlobalNotification to a seperate directory

bkoval committed:
Moving css to extension directory and renaming venus file

bkoval committed:
Fixing filepaths and whitespacing in css file for oasis

bkoval committed:
Renaming GlobalNotification to BannerNotifications

bkoval committed:
Renaming NotificationsController to BannerNotificationsController

bkoval committed:
AMD module and big cleanups in JS for BannerNotifications

bkoval committed:
Introducing BannerNotifications on Monobook

bkoval committed:
Refactoring AMD module so it returns instance of the notification

bkoval committed:
Fixing all the occurences of BannerNotifications

bkoval committed:
Adding general connection error message

bkoval committed:
Using amd module where possible

bkoval committed:
Inlining message

bkoval committed:
Changing name of a css class and fixing wikiainyourlang

bkoval committed:
Dont proceed with scroll handler if module not initialized

bkoval committed:
whitespaces

lizlux commented on extensions/wikia/BannerNotifications/BannerNotifications.setup.php:
```diff
@@ -0,0 +1,43 @@
+$wgExtensionCredits['bannernotifications'][] = array(
+	'name' => 'BannerNotifications',
+	'descriptionmsg' => 'bannernotifications-desc',
+	'author' => 'Hyun Lim',
```
Let's update this value since this extension has changed a bunch since Hyun last touched it.

bkoval replied:
I'll check the git history for this file and will update it :)

lizlux commented on extensions/wikia/BannerNotifications/css/BannerNotifications.monobook.scss:
```diff
@@ -0,0 +1,6 @@
+@import 'skins/shared/styles/buttons/buttons-wikia-chiclet';
```
Is this import necessary?

bkoval replied:
It's here so the 'close' button looks like it does on Oasis. Deleting it would make the close button look unstyled. I could add some minimal styles so it's visible enough, I just added these styles to make BN on both skins look the similar. WHat do you think, @lizlux ?

lizlux replied:
It's really not a big deal, but I wish there were a more semantic place to put that include. 

Effectively this will include the chiclet styles in all of monobook, since the BannerNotifications extension is included on every page view. This is, of course, what we want, but what happens when we remove this extension and there were other places in monobook that were relying on that chiclet styling to be there? I wish we had a `monobook.scss` like we have `oasis.scss`, that included all the core css modules needed in monobook. Maybe monobook is just too far gone to be worrying about semantic code optimizations like this though. 

Just food for thought, not a blocker as far as I'm concerned.

lizlux commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,269 @@
+	 * Bind close event to close button
+	 */
+	function setUpClose() {
+		$(document.body).on('click', '.close', onCloseClicked);
```
The selector `.close` is probably not specific enough, since there's going to be `.close` elements in places other than banner notifications, and there may be unexpected consequences if `onCloseClicked` is fired on elements like modals.

bkoval replied:
Good catch! I had in mind to change the filtering selector to '.banner-notification .close' but must've forgot to do so, thanks!

lizlux commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,269 @@
+	 * @param {String} [type]
+	 * @returns {jQuery}
+	 */
+	function createMarkup(content, type) {
```
It would be nice to create a mustache template and import it via `grunt-mustache`. https://one.wikia-inc.com/wiki/Engineering/Grunt-Mustache

bkoval replied:
Done :dancer:

lizlux commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,269 @@
+
+		// Share scroll event with WikiaFooterApp's toolbar floating (BugId:33365)
+		if (window.WikiaFooterApp) {
+			window.WikiaFooterApp.addScrollEvent();
```
Did you look into removing this dependency? I can see why they did it this way for performance reasons, but it might be nice to abstract the scroll event handling into a separate module that both `WikiaFooterApp` and `BannerNotifications` could plug into. The current dependency seems brittle and unexpected from the point of view of someone looking at this for the first time.

lizlux commented on skins/oasis/js/WikiaFooter.js:
```diff
@@ -12,7 +12,7 @@ var WikiaFooterApp = {
 		}
 		this.toolbar = this.footer.children( '.toolbar' );
-		this.gn = $( '.global-notification' );
+		this.gn = $( '.banner-notification' );
```
`gn` => `bn`

lizlux commented on skins/oasis/js/WikiHeader.js:
```diff
@@ -393,7 +393,7 @@
 						// TODO: use mustache and promise pattern along with .getMessages
 						notifications =
-							'<div class="global-notification error">' +
+							'<div class="banner-notification error">' +
```
Looks like this should be re-written to just use `BannerNotification.show(...)`

lizlux commented:
@bkoval, I checked out the code and tried out the new JS interfaces you mentioned above, there there's a few items that didn't quite work as expected: 

* When there's two notifications present, only one of them scrolls with the page. 
* Using the following syntax doesn't work, since `setType` doesn't actually return the instance 
`var notification = BannerNotifications.show('MyMessage').setType('notify');
 notification.setContent('OtherMessage').hide();`

My suggestion would be to create a JS class for BannerNotifications, so that when we need a new instance, we just call `new BannerNotification('message', 'type')` or `new BannerNotification().setMessage('message').setType('type')`. In order for that chaining to work, you have to return `this`, which is typically done by creating new instances, as mentioned above. The end result should be that any number of notifications can be added to the page, each with their own scroll handling and close events, as well as internal references to instance methods like `setType`, etc. 

I'm happy to discuss this more OO approach further, although I'm sure there's good resources in the PL office for that discussion as well.

bkoval committed:
JS fixes

bkoval committed:
Object-orientyfying banner notifications

bkoval committed:
Adding wrapper handling for notifications

bkoval committed:
Adding global scroll event handler module

bkoval committed:
Cleanups

bkoval committed:
Fixing occurrences of BannerNotification

bkoval committed:
Moving rendering view to grunt and addressing a bit of comments

bkoval committed:
Adding authors

kvas-damian commented on extensions/wikia/AssetsManager/config.php:
```diff
@@ -922,7 +925,9 @@
 		'//resources/wikia/libraries/ghostwriter/gw.min.js',
-		'//extensions/wikia/GlobalNotification/GlobalNotification.js',
+		'//extensions/wikia/BannerNotifications/js/BannerNotifications.js',
+		'//extensions/wikia/BannerNotifications/js/templates.mustache.js',
```
Can we extract those assets into group and avoid repeatition?

kvas-damian commented on extensions/wikia/AutomaticWikiAdoption/AutomaticWikiAdoptionHelper.class.php:
```diff
@@ -337,14 +337,14 @@ static function onSkinTemplateOutputPageBeforeExec(&$skin, &$tpl) {
 					'name' => 'AutomaticWikiAdoption',
 					'dismissUrl' => $wgScript . '?action=ajax&rs=AutomaticWikiAdoptionAjax&method=dismiss',
-			), NotificationsController::NOTIFICATION_CUSTOM);
+			), BannerNotificationsController::NOTIFICATION_CUSTOM );
```
Is it possible to expose the same API for JS and PHP?

kvas-damian commented on extensions/wikia/BannerNotifications/BannerNotifications.setup.php:
```diff
@@ -0,0 +1,47 @@
+// confirmations
+$wgHooks['ArticleDeleteComplete'][] = 'BannerNotificationsController::addPageDeletedConfirmation';
+$wgHooks['ArticleUndelete'][] = 'BannerNotificationsController::addPageUndeletedConfirmation';
+#$wgHooks['EditPageSuccessfulSave'][] = 'BannerNotificationsController::addSaveConfirmation'; // BugId:10129
```
What do you think about removing this line? I don't expect it get used again when we introduced VE.

bkoval committed:
Adding group for js files and newlines at the end of files;

kvas-damian commented on extensions/wikia/FacebookClient/scripts/preferences.js:
```diff
@@ -54,7 +58,10 @@
 
-						window.GlobalNotification.show($.msg('fbconnect-preferences-connected'), 'confirm');
+						bannerNotification
+							.setContent($.msg('fbconnect-preferences-connected'))
```
I know this code was already there, but we could introduce escaping messages: http://www.mediawiki.org/wiki/Manual:Messages_API#Use_of_the_messages

bkoval replied:
@kvas-damian What do you think about creating a ticket to do so and taking care of escaping messages problem on a separate basis?

kvas-damian replied:
Sounds good for me.

kvas-damian commented on extensions/wikia/BannerNotifications/BannerNotificationsController.class.php:
```diff
@@ -5,7 +5,7 @@
  */
 
-class NotificationsController extends WikiaController {
+class BannerNotificationsController extends WikiaController {
```
I'm not sure if this class should be renamed. It's also responsible for displaying Oasis notification: https://one.wikia-inc.com/wiki/Oasis/Notifications_and_confirmations

lizlux replied:
We'll have to warn QA that the oasis notifications will need a pass since they are affected.

lizlux commented on skins/oasis/js/WikiaFooter.js:
```diff
@@ -30,7 +33,7 @@ var WikiaFooterApp = {
 		'use strict';
-		WikiaFooterApp.windowObj.off( 'scroll.FooterAp' ); // GlobalNotifications could be re-binding this event.
+		WikiaFooterApp.windowObj.off( 'scroll.FooterAp' );
 		WikiaFooterApp.windowObj.on( 'scroll.FooterAp', WikiaFooterApp.resolvePosition ).triggerHandler( 'scroll' );
```
Let's create a follow-up ticket to have WikiaFooter use the nifty new onScroll module (if one hasn't been created already).

bkoval replied:
https://wikia-inc.atlassian.net/browse/SOC-401

lizlux commented:
@bkoval, for some reason I keep getting `Uncaught Module BannerNotification is not defined.` when I run this code in a browser. The error is coming from the AMD code so I'm not quite sure why it's happening. To reproduce, you can try logging out and logging back in again.

lizlux commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,251 @@
+			self = this;
+			setTimeout(function () {
+				self.hide();
+			}, this.timeout);
```
You could write this as: 

```
setTimeout(function () {
	this.hide();
}.bind(this), this.timeout);
```
to remove the need for the local var `self`.

lizlux commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,251 @@
+		return this;
+	};
+
+	BannerNotification.prototype.showConnectionError = function () {
```
Let's add more flushed out doc blocks for the file. Also, it would be nice to order the functions a bit better too. For example, you could put the private functions below all the prototype methods, and place each method in the order in which it might be called.

lizlux commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,251 @@
+		return this;
+	};
+
+	BannerNotification.prototype.hide = function (callback) {
```
What do you think about getting rid of this `callback` param? I think it was only added to be used internally, and it looks like it's no longer being used anywhere. It would be nice to simplify the code, and it can always be added back in later if we need it in the future.

bkoval replied:
Let's do so, it's not used anywhere (at least by my investigation, which could have ommited something) :)

lizlux commented on extensions/wikia/WikiaStyleGuide/templates/WikiaStyleGuideSpecial_index.php:
```diff
@@ -250,29 +250,29 @@
+		BannerNotifications.show('Hello there, I am Banner Notification', 'notify');
 	}, function() {
-		GlobalNotification.hide();
+		BannerNotifications.hideAll();
```
Does this method exist?

lizlux commented on extensions/wikia/WikiaInYourLang/modules/ext.wikiaInYourLang.js:
```diff
@@ -101,24 +102,23 @@ require(
-			$('.global-notification.notify')
+		function bindEvents($element) {
+			$element
 				.on('click', '.close', function() {
```
This might be better handled as an `onClose` event that's emitted from the BannerNotification instance. That way we would be free to change stuff like the `.close` element without breaking stuff like this. 

Alternately, we could expose the `.close` element as `notification.$close` and bind to that.

bkoval replied:
Went with the first approach.

lizlux commented on skins/oasis/css/core/responsive-core.scss:
```diff
@@ -230,7 +230,7 @@
 
-		.global-notification {
+		.banner-notification {
 			width: 100%;
```
This is causing some funky behavior on small screens. 
![screen shot 2015-02-19 at 6 10 02 pm](https://cloud.githubusercontent.com/assets/201367/6280488/ed9bbc0e-b862-11e4-922f-e261b74b292b.png)

lizlux commented on skins/oasis/js/WikiHeader.js:
```diff
@@ -392,12 +391,12 @@
 					if ( errorMessages.length > 0 ) {
 						$( '#publish' ).remove();
 						// TODO: use mustache and promise pattern along with .getMessages
```
This comment should be updated.

lizlux commented on skins/oasis/js/WikiaFooter.js:
```diff
@@ -12,12 +15,12 @@ var WikiaFooterApp = {
 		}
 		this.toolbar = this.footer.children( '.toolbar' );
-		this.gn = $( '.global-notification' );
+		this.bn = $( '.banner-notification' );
```
Actually it looks like we can remove references to banner notifications from this file all together, since we don't need the scroll event or the resize event for BN.

lizlux commented on extensions/wikia/WikiaStyleGuide/templates/WikiaStyleGuideSpecial_index.php:
```diff
@@ -250,29 +250,29 @@
+	BannerNotifications.show('Hello there, I am Banner Notification', 'notify');
 }, function() {
-	GlobalNotification.hide();
+	BannerNotifications.hideAll();
```
Same

bkoval committed:
CR changes

kvas-damian commented on extensions/wikia/AssetsManager/config.php:
```diff
@@ -922,7 +924,8 @@
 		'//resources/wikia/libraries/ghostwriter/gw.min.js',
-		'//extensions/wikia/GlobalNotification/GlobalNotification.js',
+		'#group_banner_notifications_js',
+		'//skins/shared/scripts/onScroll.js',
```
```#group_banner_notifications_js``` is dependent on ```//skins/shared/scripts/onScroll.js```. Don't you think we should solve this dependancy somehow?

bkoval replied:
After offline discussion with @kvas-damian , we decided to keep things the way there are and consider resolving dependency management for AssetManager as a seperate cause.

kvas-damian commented on extensions/wikia/BannerNotifications/templates/BannerNotifications.mustache:
```diff
@@ -0,0 +1,6 @@
+<div class="banner-notification">
+	<button class="close wikia-chiclet-button">
+		<img src="{{imageSource}}">
```
Is it possible to render this image using CSS?
This way we could limit 1 request for assets and make JS code cleaner.

bkoval replied:
Done.

kvas-damian commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,294 @@
+	function createBackendNotification() {
+		var $backendNotification = $('.banner-notification');
+		if ($backendNotification.length) {
+			backendNotification = new BannerNotification();
```
Don't we need to pass at least ```type``` and ```content``` arguments?

kvas-damian replied:
Do we support multiple instances of ```backendNofications```?

kvas-damian commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,294 @@
+		$(notification.$element).on('click', '.close', function (event) {
+			notification.hide();
+			if (notification.onCloseHandler) {
+				notification.onCloseHandler(event);
```
what do you think about defining ```onClonseHandler``` as ```Function.prototype``` by default and just call it without checking if it's defined.

kvas-damian commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,294 @@
+	 */
+	function removeFromDOM($elements) {
+		if ($elements.length) {
+			$elements.animate({
```
Do you think this animation should be part of ```removeFromDOM``` method?

kvas-damian commented on skins/shared/scripts/onScroll.js:
```diff
@@ -0,0 +1,54 @@
+	 * Binds global scroll event listener
+	 */
+	function init() {
+		window.addEventListener('scroll', trigger);
```
Don't you think we need at least small debounce for better performance?

bkoval replied:
Let's add a minimal one.

bkoval committed:
Fixing layout problems

bkoval committed:
CSS improvements and getting rid of img src

bkoval committed:
Require module to trigger it on the fly and extracting animation

bkoval committed:
Using onClose in WikiaInYourLang and fixing name in cookiePolicy

bkoval committed:
Making separate controllers for banner and oasis notifications

bkoval committed:
Commenting out test code again

bkoval commented:
Few changes:
 - To avoid the undefined module problem, I wrapped the module into a 'require' block that exports an AMD module at the very end, to trigger the code right away
 - Divided controllers for BannerNotifications and Oasis Notifications
 - Added 'onClose' handler
 - Added little debounce (5ms) to onScroll module

bkoval committed:
Fixing comments

bkoval committed:
Moving wikiainyourlang to assetsman to remove race condition with bn

kvas-damian commented on extensions/wikia/BannerNotifications/css/BannerNotifications.scss:
```diff
@@ -102,8 +83,36 @@
 		img {
-			height: 11px;
-			width: 12px;
+			background-image: url('/skins/oasis/images/icon_close.png');
```
Please use inline image to avoid additional request.

kvas-damian commented on extensions/wikia/BannerNotifications/BannerNotificationsController.class.php:
```diff
@@ -0,0 +1,292 @@
+	/**
+	 * Add confirmation message to the user session (so it persists between redirects)
+	 */
+	public static function addConfirmation($message, $type = 1) {
```
please use constant for ```$type```

kvas-damian commented on extensions/wikia/BannerNotifications/BannerNotificationsController.class.php:
```diff
@@ -0,0 +1,292 @@
+		#wfRunHooks('SkinTemplatePageBeforeUserMsg', array(&$ntl));
+
+		// add testing confirmation
+		// self::addConfirmation('test');
```
I don't think we need debugging code here.

kvas-damian commented on extensions/wikia/BannerNotifications/BannerNotificationsController.class.php:
```diff
@@ -0,0 +1,292 @@
+		// add testing confirmation
+		// self::addConfirmation('test');
+
+		if (!empty($_SESSION[self::SESSION_KEY])) {
```
whitespaces ;)

kvas-damian commented on extensions/wikia/BannerNotifications/BannerNotificationsController.class.php:
```diff
@@ -0,0 +1,292 @@
+
+			switch($entry['type']) {
+				case self::CONFIRMATION_NOTIFY:
+					$this->confirmationClass = 'notify';
```
what do you think about storing ```$type``` as a string value in session?
We could avoid this mapping then.

kvas-damian commented on extensions/wikia/BannerNotifications/BannerNotificationsController.class.php:
```diff
@@ -0,0 +1,292 @@
+
+		if ( F::app()->checkSkin( 'oasis' ) ) {
+			if ($wgRequest->getCheck('success')) {
+				self::addConfirmation(wfMsg('savedprefs'));
```
Please use new ```wfMessage``` instead of deprecated ```wfMsg``` and escape strings.

kvas-damian commented on extensions/wikia/BannerNotifications/BannerNotificationsController.class.php:
```diff
@@ -0,0 +1,292 @@
+	 *
+	 * @return true
+	 */
+	function onOutputPageBeforeHTML( &$out, &$text ) {
```
please use full method signature - including visibility and static keyword

bkoval replied:
static public it is! ;)

kvas-damian commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,306 @@
+	BannerNotification.prototype.showConnectionError = function () {
+		return this
+			.setType('error')
+			.setContent($.msg('bannernotifications-general-ajax-failure'))
```
Please use escaped message ;)

bkoval committed:
Addressing Damians notes

bkoval committed:
Loading message through resource loader and escaping it

bkoval committed:
Removing development leftovers

bkoval committed:
Removing development leftovers

lizlux commented:
@bkoval, I tried to check out the code and run it, but I'm seeing escaped HTML as the output in the browser. See screenshot for error message in page source: 

![screen shot 2015-02-26 at 3 10 25 pm](https://cloud.githubusercontent.com/assets/201367/6404143/adf93038-bdc9-11e4-82fb-51f4117e8dfc.png)

lizlux commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,309 @@
+	}
+
+	/**
+	 * Creates an instance of a notification set on the backend side
```
This message seems unclear to someone that doesn't know about notifications originating from the server. Maybe update the wording to something like: 

"Create instance of BannerNotification based on HTML passed from the server on page load."

bkoval commented:
@lizlux Could you try to checkout this branch again? I guess this is connected to the Mac's filesystem being case-insensitive, I forced the changes in git to use the capital letter now :)

lizlux commented:
@bkoval thanks for the quick fix!

lizlux commented:
I'm still seeing that JS error that the module BannerNotification isn't defined, right at page load. This is definitely a race condition tho b/c I'm consistently seeing it being defined after the page has fully loaded. 

![screen shot 2015-02-26 at 3 35 44 pm](https://cloud.githubusercontent.com/assets/201367/6404473/5ddc0a04-bdcd-11e4-87d6-e3e3e1e09cff.png)

Saw the JS error here: http://lizlux.liz.wikia-dev.com/wiki/TabViewContent (just a random article on my wikia).

lizlux commented:
I must say the functionality is looking really good, and the API is really smooth. It's a pleasure to play around with it. 
:+1:

lizlux commented on extensions/wikia/WikiaInYourLang/WikiaInYourLang.hooks.php:
```diff
@@ -9,12 +9,12 @@
 	 */
 	public static function onBeforePageDisplay( \OutputPage $out ) {
-		$out->addModules( 'ext.wikiaInYourLang' );
+		\Wikia::addAssetsToOutput( 'wikia_in_your_lang_js' );
```
Is this how we're supposed to be doing this now? I feel like there's 100 ways to do add assets. :grimacing:

bkoval replied:
I like this way of adding assets, this makes it independent from the hook it is used from. But I'm open to suggestions if you feel there's a better way :)

lizlux commented on skins/oasis/css/core/responsive-core.scss:
```diff
@@ -13,19 +13,23 @@
 
-	.global-notification {
+	.banner-notification {
+		@include box-sizing(border-box);
```
We don't need a mixin for this anymore. `box-sizing: border-box` works on all supported browsers now. :)
http://caniuse.com/#search=box-sizing

Someone should really clean that up...

kvas-damian replied:
Lets do that: https://wikia-inc.atlassian.net/browse/SOC-445

lizlux commented on skins/oasis/modules/NotificationsController.class.php:
```diff
@@ -7,14 +7,6 @@
 
 class NotificationsController extends WikiaController {
 
-	const SESSION_KEY = 'oasisConfirmation';
```
(Not related to this line, but) should the doc block on line 3 be updated to remove the reference to "(green stripe below global nav)"? It looks like you split this file out into the two types of notifications (which is good).

lizlux commented on extensions/wikia/BannerNotifications/BannerNotifications.setup.php:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Hooks
+ */
+$wgHooks['ArticleDeleteComplete'][] = 'BannerNotificationsController::addPageDeletedConfirmation';
```
It would be good to mention to QA that these are some back end instances to test, since it's less well know that these can come from the back end.

bkoval replied:
Sure thing, I'll prepare a description for QA-ing this :)

lizlux commented on extensions/wikia/BannerNotifications/BannerNotificationsController.class.php:
```diff
@@ -0,0 +1,291 @@
+
+class BannerNotificationsController extends WikiaController {
+
+	const SESSION_KEY = 'oasisConfirmation';
```
Should we update this to `BannerNotification`?

lizlux commented on extensions/wikia/BannerNotifications/BannerNotificationsController.class.php:
```diff
@@ -0,0 +1,291 @@
+			$this->confirmationClass = $entry['type'];
+
+			// clear confirmation stack
+			unset( $_SESSION[self::SESSION_KEY] );
```
Looks like `clearConfirmation` is never being called. Let's either delete it or call it from here.

lizlux commented on extensions/wikia/BannerNotifications/BannerNotificationsController.class.php:
```diff
@@ -0,0 +1,291 @@
+			wfDebug( __METHOD__ . " - {$message}\n" );
+		}
+
+		wfProfileOut( __METHOD__ );
```
Let's rip these `wfProfileOut` `wfProfileIn` calls out since we don't need them anymore.

lizlux commented on extensions/wikia/FacebookClient/FacebookClientHooks.class.php:
```diff
@@ -127,7 +127,7 @@ public static function onSkinTemplatePageBeforeUserMsg( &$html ) {
 				$map = FacebookClient::getInstance()->getMapping();
 				if ( !empty( $map ) ) {
-					NotificationsController::addConfirmation( wfMessage( 'fbconnect-connect-msg' )->plain() );
+					BannerNotificationsController::addConfirmation( wfMessage( 'fbconnect-connect-msg' )->plain() );
```
`escaped()`

lizlux commented on extensions/wikia/SpecialCss/SpecialCssController.class.php:
```diff
@@ -41,29 +41,29 @@ public function index() {
 				if (!$status) {
-					NotificationsController::addConfirmation(
+					BannerNotificationsController::addConfirmation(
 						wfMessage('special-css-merge-error')->plain(),
```
escape

lizlux commented on extensions/wikia/SpecialCss/SpecialCssController.class.php:
```diff
@@ -41,29 +41,29 @@ public function index() {
 					$this->diff = $this->app->sendRequest(__CLASS__, 'getDiff', ['wikitext' => $content])->getVal('diff');
 				} else if ($status->isOk()) {
-					NotificationsController::addConfirmation( wfMessage('special-css-save-message')->plain() );
+					BannerNotificationsController::addConfirmation( wfMessage('special-css-save-message')->plain() );
```
escape

lizlux commented on extensions/wikia/SpecialCss/SpecialCssController.class.php:
```diff
@@ -41,29 +41,29 @@ public function index() {
 		if ($this->request->getVal('oldid', null) !== null) {
-			NotificationsController::addConfirmation(
+			BannerNotificationsController::addConfirmation(
 				wfMessage('special-css-oldid-message')->plain(),
```
escape (obviously not your code but would be good to update)

lizlux commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,309 @@
+ *
+ * Use like:
+ * new BannerNotification('Content', 'error').show().hide();
+ */
```
Might be worth noting from a security perspective that this module doesn't escape content before inserting it into the DOM, so it's only to be used with content that's already considered safe (i.e. no UCG).

lizlux commented on extensions/wikia/BannerNotifications/js/BannerNotifications.js:
```diff
@@ -0,0 +1,309 @@
+ * Use like:
+ * new BannerNotification('Content', 'error').show().hide();
+ */
+require([
```
Also, let's add some unit tests! (Can be done in a separate ticket)

bkoval commented:
Hey @lizlux ! Thanks for checking out the branch and for CR. This weird race condition worries me though. I was trying to reproduce it using SF devbox (dev-interactivemaps) and was wandering around different wikias for quite some time. I didn't see the race condition error though, all was loading fine:

https://www.dropbox.com/s/3sn4kvxe60yfuy4/Zrzut%20ekranu%202015-02-27%2011.41.25.png?dl=0

Not sure where does it come from, but I'll set up a Sandbox today and try to reproduce it as well.

bkoval committed:
Addressing more comments

bkoval commented:
https://wikia-inc.atlassian.net/browse/SOC-446 - ticket for adding unit tests.

bkoval committed:
Removing profiling

kvas-damian commented:
Besides error mentioned by @lizlux it looks ok for me.
I tried to find it but with no luck. Liz are you still able to reproduce this issue? 
If yes, it'd be helpful if you provide us more info, like browser etc.
